src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,395 @@
+            except Exception as e:
+                # TODO: deal with this (log it? ignore it?)
+                # e.g. RADOS rados state (You cannot perform that operation on a Rados object in state configuring.)
+                # this makes the log pretty messy prior to bootstrap (errors every 10 seconds)
+                logger.error(str(e))
+
+            # TODO: Should we sleep for longer?  A minute instead of 10 seconds?  This is pretty arbitrary...
+            logger.debug("Cluster connection thread sleeping for 10 seconds")
+            self._event.wait(10)
+            self._event.clear()
+            # How on earth to we detect that the cluster has gone away
+            # and trigger a reconnect?
+            # What happens if we try a write op on a down cluster?
+            # might want watch checks?
+
+        logger.debug("Shutting down cluster connection")
+        if self._config_watch:
+            self._config_watch.close()
+            # Need to set this to None, so it's deallocated before the
+            # cluster is deallocated/shutdown, or we get:
+            # Traceback (most recent call last):
+            #   File "rados.pyx", line 477, in rados.Rados.require_state
+            # rados.RadosStateError: RADOS rados state (You cannot perform that operation on a Rados object in state shutdown.)
+            # Exception ignored in: 'rados.Watch.__dealloc__'
+            # Traceback (most recent call last):
+            #   File "rados.pyx", line 477, in rados.Rados.require_state
+            # rados.RadosStateError: RADOS rados state (You cannot perform that operation on a Rados object in state shutdown.)
+            self._config_watch = None
+        if self._ioctx:
+            self._ioctx.close()
+        if self._cluster:
+            self._cluster.shutdown()
+        logger.debug("Cluster connection is shut down")
+
+    def _config_notify(
+        self, notify_id: int, notifier_id: str, watch_id: int, data: bytes
+    ) -> None:
+        key = data.decode("utf-8")
+        logger.debug(
+            f"Got notify on config object {notify_id} {notifier_id} {watch_id} {key}"
+        )
+        if key not in self._watches:
+            return
+
+        value = self._get(key)
+        for watch in self._watches[key].values():
+            watch(key, value)
 
     async def close(self) -> None:
         """Close k/v store connection"""
-        self._is_closing = True
-        if not self._client:
-            self._is_open = False
-            return
-        await self._client.close()
-        self._client = None
-        self._is_open = False
+        self._run = False
+        self._event.set()
 
     async def put(self, key: str, value: str) -> None:
         """Put key/value pair"""
-        assert self._client
-        await self._client.put(key, value)
+        # Try to write to the kvstore in our pool, and also
+        # write to our local cache (whether or not the write
+        # to the pool succeeds)
```

Comment:
I think we should only write to the cache if the write succeeds, and ideally from the watches. If we're not watching for a specific key, I think it's okay to not cache it. This way we ensure that we always have a consistent state with the cluster, even if it gets out of sync when we die.

Reply:
Perhaps those need to be tracked in a separate cache so not to confuse ourselves with the state of the cluster?

But I do see the below comments on getting back into sync being very difficult.